src/Control/Monad/Metrics/Internal.hs

```diff
@@ -25,46 +36,114 @@ import           System.Metrics.Distribution (Distribution)
 import           System.Metrics.Gauge        (Gauge)
 import           System.Metrics.Label        (Label)
 
+type IdenticalKeyTypes m m' =
+    ( CounterKey m      ~ CounterKey m'
+    , GaugeKey m        ~ GaugeKey m'
+    , DistributionKey m ~ DistributionKey m'
+    , LabelKey m        ~ LabelKey m'
+    )
+
+type ValidMetricKeys m =
+    ( MetricKey (CounterKey m)
+    , MetricKey (GaugeKey m)
+    , MetricKey (DistributionKey m)
+    , MetricKey (LabelKey m)
+    )
+
+type CounterKeyFor      m key = (MetricKey key, key ~ CounterKey m)
+type GaugeKeyFor        m key = (MetricKey key, key ~ GaugeKey m)
+type DistributionKeyFor m key = (MetricKey key, key ~ DistributionKey m)
+type LabelKeyFor        m key = (MetricKey key, key ~ LabelKey m)
+
+-- | We can use arbitrary key types for our 'MonadMetrics'es, but to do so we
+-- need to a way to look the types up. Sometimes we want this association even
+-- for a type that can't be a 'MonadMetrics', like 'IO'.
+--
+-- 'Meterable' provides that capability. It has a 'MonadTrans' instance, so
+-- transformer stacks will use the key types for the monad they build upon.
+class Monad m => Meterable m where
+    type CounterKey      m :: *
+    type GaugeKey        m :: *
+    type DistributionKey m :: *
+    type LabelKey        m :: *
```

Comment:
We can provide defaults here, too, so you can just write `instance Meterable MyMonad` and it gets `Text` for all of them.

Reply:
What makes that a better approach than making all associated types explicit?

Reply:
Backwards compatibility/convenience, mostly. Right now this change is going to require users to provide an `instance Meterable TheirAppType where type CounterKey = Text; -- etc`. Providing defaults makes the migration easier.

I'm actually confused now why this is a separate class. I remember you split it off but not why :\

Reply:
If there are a chain of `MonadTrans` instances down to the base type, it's not going to require users to provide a `Meterable` instance.

This is why it's a separate class: https://github.com/parsonsmatt/monad-metrics/commit/ddbdeed